File: web-mobile/pagamento-semplice/pagamento-multivaluta/python/esito.py
# -*- coding: utf-8 -*-
import sys
if sys.version_info >= (3,):
    from urllib.parse import urlencode
else:
    from urllib import urlencode
import hashlib
import datetime

param_from_request = {
"codTrans":"",
"esito":"OK",
"importo":"",
"divisa":"",
"data":"",
"orario":"",
"codAut":"",
"mac":"58815356e9052f7d6a355a399d1d8edfc4a58bd7"
}

# Chiave segreta
CHIAVESEGRETA = "<CHIAVE SEGRETA PER CALCOLO MAC>" # Sostituire con il valore fornito da CartaSi

# Controllo che ci siano tutti i parametri di ritorno obbligatori per calcolare il MAC
requiredParams = ['codTrans', 'esito', 'importo', 'divisa', 'data', 'orario', 'codAut', 'mac']
for param in requiredParams:
    if param not in param_from_request:
        raise ValueError("Parametro {} mancante".format(param))

# Calcolo MAC con i parametri di ritorno
mac_str = 'codTrans=' + param_from_request['codTrans'] + \
        'esito=' + param_from_request['esito'] + \
        'importo=' + param_from_request['importo'] + \
        'divisa=' + param_from_request['divisa'] + \
        'data=' + param_from_request['data'] + \
        'orario=' + param_from_request['orario'] + \
        'codAut=' + param_from_request['codAut'] + \
        CHIAVESEGRETA
macCalculated =  hashlib.sha1(mac_str.encode('utf8')).hexdigest()

# Verifico corrispondeza tra MAC calcolato e parametro mac di ritorno
macCalculated != param_from_request['mac']:
    raise ValueError('Errore MAC: ' + macCalculated +' non corrisponde a ' + param_from_request['mac'])

# Nel caso in cui non ci siano errori gestisco il parametro esito
if param_from_request['esito'] == 'OK':
    print('La transazione ' + param_from_request['codTrans'] + " è avvenuta con successo; codice autorizzazione: " + param_from_request['codAut'])
    print('dccRate = ' + param_from_request['dccRate'])
    print('dccAmount = ' + param_from_request['dccAmount'])
    print('dccCurrency = ' + param_from_request['dccCurrency'])
    print('dccState = ' + param_from_request['dccState'])
else:
    print('La transazione ' + param_from_request['codTrans'] + " è stata rifiutata")
    if 'messaggio' in param_from_request:
        print("descrizione errore: " + param_from_request['messaggio'])

File: web-mobile/pagamento-semplice/pagamento-multivaluta/php/esito.php
<?php

// Chiave segreta 
$CHIAVESEGRETA = "<CHIAVE SEGRETA PER CALCOLO MAC>"; // Sostituire con il valore fornito da Nexi

// Controllo che ci siano tutti i parametri di ritorno obbligatori per calcolare il MAC
$requiredParams = array('codTrans', 'esito', 'importo', 'divisa', 'data', 'orario', 'codAut', 'mac');
foreach ($requiredParams as $param) {
    if (!isset($_REQUEST[$param])) {
        echo 'Paramentro mancante ' . $field;
        exit;
    }
}

// Calcolo MAC con i parametri di ritorno
$macCalculated = sha1('codTrans=' . $_REQUEST['codTrans'] .
        'esito=' . $_REQUEST['esito'] .
        'importo=' . $_REQUEST['importo'] .
        'divisa=' . $_REQUEST['divisa'] .
        'data=' . $_REQUEST['data'] .
        'orario=' . $_REQUEST['orario'] .
        'codAut=' . $_REQUEST['codAut'] .
        $CHIAVESEGRETA
);

// Verifico corrispondenza tra MAC calcolato e parametro mac di ritorno
if ($macCalculated != $_REQUEST['mac']) {
    echo 'Errore MAC: ' . $macCalculated . ' non corrisponde a ' . $_REQUEST['mac'];
    exit;
}

// Nel caso in cui non ci siano errori gestisco il parametro esito
if($_REQUEST['esito'] == 'OK'){
    echo 'La transazione ' . $_REQUEST['codTrans'] . " è avvenuta con successo; codice autorizzazione: " . $_REQUEST['codAut'];
    echo    '<br><br><b>DDC</b>'
            . '<br>dccRate = ' . $_REQUEST['dccRate']
            . '<br>dccAmount = ' . $_REQUEST['dccAmount']
            . '<br>dccCurrency = ' . $_REQUEST['dccCurrency']
            . '<br>dccState = ' . $_REQUEST['dccState'];
} else {
    echo 'La transazione ' . $_REQUEST['codTrans'] . " è stata rifiutata; descrizione errore: " . $_REQUEST['messaggio'];
}

